AppCenter/AppCenter/Internals/Model/MSAbstractLog.m

```diff
@@ -119,4 +117,101 @@ - (void)addTransmissionTargetToken:(NSString *)token {
   }
 }
 
+#pragma mark - MSLogConversion
+
+- (NSArray<MSCommonSchemaLog *> *)toCommonSchemaLogs {
+  NSMutableArray<MSCommonSchemaLog *> *csLogs = [NSMutableArray new];
+  for (NSString *token in self.transmissionTargetTokens) {
+    MSCommonSchemaLog *csLog = [self toCommonSchemaLogForTargetToken:token];
+    if (csLog) {
+      [csLogs addObject:csLog];
+    }
+  }
+
+  // Return nil if none are converted.
+  return (csLogs.count > 0) ? csLogs : nil;
+}
+
+#pragma mark - Helper
+
+- (MSCommonSchemaLog *)toCommonSchemaLogForTargetToken:(NSString *)token {
+  MSCommonSchemaLog *csLog = [MSCommonSchemaLog new];
+  csLog.ver = kMSCSVerValue;
+  csLog.timestamp = self.timestamp;
+
+  // TODO popSample not supported at this time.
+
+  // Calculate iKey based on the target token.
+  if (token && token.length) {
+    csLog.iKey = [NSString stringWithFormat:@"o:%@", [token componentsSeparatedByString:@"-"][0]];
+  }
+
+  // TODO flags not supported at this time.
+  // TODO cV not supported at this time.
+
+  // Setup extensions.
+  csLog.ext = [MSCSExtensions new];
+
+  // Protocol extension.
+  csLog.ext.protocolExt = [MSProtocolExtension new];
+  csLog.ext.protocolExt.devMake = self.device.oemName;
+  csLog.ext.protocolExt.devModel = self.device.model;
+
+  // User extension.
+  csLog.ext.userExt = [MSUserExtension new];
+
+  // Convert user local to use dash (-) as the separator as described in RFC 4646.  E.g., zh-Hans-CN.
+  csLog.ext.userExt.locale = [self.device.locale stringByReplacingOccurrencesOfString:@"_" withString:@"-"];
+
+  // OS extension.
+  csLog.ext.osExt = [MSOSExtension new];
+  csLog.ext.osExt.name = self.device.osName;
+  csLog.ext.osExt.ver = [self combineOsVersion:self.device.osVersion withBuild:self.device.osBuild];
+
+  // App extension.
+  csLog.ext.appExt = [MSAppExtension new];
+  csLog.ext.appExt.appId = [NSString stringWithFormat:@"I:%@", self.device.appNamespace];
+  csLog.ext.appExt.ver = self.device.appVersion;
+
+  // TODO app.locale not supported at this time.
```

Comment:
well on iOS the locale you collect is not user but app one right?

Reply:
Right because it's limited to the locale "the app" supports, so we should move that to the app locale.

Reply:
Actually it needs some more testing because it competes with another candidate to get the app local which is [[NSBundle mainBundle] preferredLocalizations]. Will do more testing to see what best fits the app locale.